src/mod/server.mod/servmsg.c

```diff
@@ -1148,6 +1266,138 @@ static int got465(char *from, char *msg)
+  int len, i = 0;
+
+  newsplit(&msg);
+  putlog(LOG_DEBUG, "*", "CAP: %s", msg);
+  cmd = newsplit(&msg);
+  fixcolon(msg);
+  if (!strcmp(cmd, "LS")) {
+    putlog(LOG_DEBUG, "*", "CAP: %s supports CAP sub-commands: %s", from, msg);
+    strlcpy(cap.supported, msg, sizeof cap.supported);
+    create_cap_req();
+    /* Check each desired cape against supported capes on server */
+    for (i = 0; i < (sizeof capes / sizeof capes[0]); i++) {
+      if (strlen(capes[i])) {
+        if (strstr(cap.supported, capes[i])) {
+          strncat(cap.desired, capes[i], (CAPMAX - strlen(cap.desired) - 1));
+          strncat(cap.desired, " ", (CAPMAX - strlen(cap.desired) - 1));
+        } else {
+          putlog(LOG_DEBUG, "*","CAP: desired capability %s not supported " 
+              "on %s, removing from request...", capes[i], from);
+        }
+      }
+    }
+    putlog(LOG_DEBUG, "*", "CAP: Requesting %s capabilities from server", cap.desired);
+    if (strlen(cap.desired) > 0) {
+      dprintf(DP_MODE, "CAP REQ :%s", cap.desired);
+    } else {
+      dprintf(DP_MODE, "CAP END");
+    }
+  } else if (!strcmp(cmd, "LIST")) {
+    putlog(LOG_SERV, "*", "CAP: Negotiated CAP capabilities: %s", msg);
+    strlcpy(cap.negotiated, msg, sizeof cap.negotiated);
+  } else if (!strcmp(cmd, "ACK")) {
+    if (msg[0] == '-') {
+      msg++;
+      len = strlen(msg);    /* Remove capability from .negotiated list */
+      while ((match = strstr(cap.negotiated, msg))) {
+        *match = '\0';
+        strcat(cap.negotiated, match+len);
+      }
+      putlog (LOG_SERV, "*", "CAP: Disabled %s with %s", msg, from);
+    } else {
+      putlog(LOG_SERV, "*", "CAP: Successfully negotiated %s with %s", msg, from);
+      if (!strstr(cap.negotiated, msg)) {  //TODO break apart msg in case multiple capes requeseted at once
+        strncat(cap.negotiated, msg, (sizeof cap.negotiated -
+            strlen(cap.negotiated) - 1));
+      }
+      /* If a negotiated capability requires immediate action by Eggdrop,
+       * add it here                                                   */
+      if (strstr(msg, "sasl") != NULL) {
+        if (sasl_mechanism[0]) {
+          for (i = 0; i < strlen(sasl_mechanism); i++)
+            sasl_mechanism[i] = toupper(sasl_mechanism[i]);
+          if (HAVE_OPENSSL_SSL_H || strncmp(sasl_mechanism,
+              "ECDSA-NIST256P-CHALLENGE", strlen("ECDSA-NIST256P-CHALLENGE"))) {
+            /*
+            TODO: the old sasl code, before cap pr, was doing cap request only
+            under certain conditions, see the if HAVE_OPENSSL_SSL_H statement
+            above.
+            putlog(LOG_SERV, "*", "CAP: put CAP REQ :sasl");
+            dprintf(DP_MODE, "CAP REQ :sasl\n");
+            */
+            putlog(LOG_SERV, "*", "SASL: put AUTHENTICATE %s", sasl_mechanism);
+            dprintf(DP_MODE, "AUTHENTICATE %s\n", sasl_mechanism);
+          }
+        }
+      }
+    }
+    dprintf(DP_MODE, "CAP END");
```

Comment:
This is getting sent before the server can reply to finish SASL, thus terminating the negotiation prior to completion and causing an Abort message to be generated.

Maybe the solution lies in a timeout functionality? If no capes requested need negotiation during registration, send the end immediately, otherwise wait up to X seconds for a server response then send the END and move on?

Reply:
in my test, sasl plain and sasl ecdsa is working fine. but yes, i will analyze this issue later and leave it as TODO now.